### ISSUE TITLE
Using a sysroot

### ISSUE THREAD

**Issue by Ignition** [conan-io/conan#5604]
I have the need to make builds for other sysroots, so I've packaged the compiler with a new setting to distinguish between sysroot. To use it I do the following:
```
USER_CHANNEL="third_party/stable"

conan install gcc_installer/9.1.0@build_tools/stable -g=virtualenv \
  -s arch_target=x86_64 \
  -s os_target=Linux \
  -s sysroot_target=rhel7

rm -f ./profile_to_use
source activate.sh
conan profile new ./profile_to_use --detect
conan profile update settings.compiler.libcxx=libstdc++11 ./profile_to_use
conan profile update settings.sysroot=rhel7 ./profile_to_use
conan create . ${USER_CHANNEL} -pr profile_to_use
conan search mylib/1.0.0@${USER_CHANNEL}
source deactivate.sh
```

It is not a good as I want. I am looking for feedback on refinement on the approach.

1) `sysroot_target` and `sysroot` are new and I had to add them to `~/.conan/settings.yml`. Is this the best setting? An alternative could be `platform`?
2) I have to define a new profile that matches the compiler. Should `virtualenv` be expanded to do this?
3) I have to pass in that profile into `conan create`. `gcc_installer` could package its own profile which could be export by an environmental variable. Should conan set a profile based of an environmental variable?
4) mylib must have `sysroot` in its settings so that the package id is distinct. I've forgotten a few times. Any idea on was of mitigating this accident?

**Comment by memsharded** [comment]
Hi @Ignition 

I would like to understand a bit better about your goals and what are you trying to achieve. Usually, for modelling distros that might have different ABI compatibility (the most typical one is RH6 vs RH7) the most common approach is adding a subsetting to ``os: Linux`` setting, not a new root one. In that way, this is something that only applies for Linux distros, and doesn't affect at all to all other OSs. ``None`` is a valid value, if you want to model the "generic" or no specific distro. Also, subsettings are automatically injected, no need to modify recipes to add them, as long as they are defined in the profiles, the package-ID will be distinct.

How are you using that ``sysroot`` values in the recipes?

**Comment by Ignition** [comment]
This is completely an ABI compatibility thing. As a developer I work on Ubuntu, but our release is for currently RHEL6, soon RHEL7. I need to make sure those packages are separate to ensure correctly released product. I've also invested heavily into using cross compilers to remove the need of having dedicated build machines. 

I agree subsetting makes more sense. I've changed to a `distro` subsetting as it is a more natural fit.

Having said that I'm having issues with using subsetting for `os_target` since all packages need to be the same. `gcc_installer/9.1.0@build_tools/stable` is sensitive to the distro as it needs to provide `libgcc_s`, `libstdc++`, etc for the target sysroot. But `binutils_installer/2.32@build_tools/stable` shouldn't care, but it does. With the subsetting I'd have to build two packages which are identical, just to satisfy the new subsetting. Current fix is to use an option rather than `os_target.distro`.

I also have `sysroot_installer/2.0.0@build_tools/stable` which has packaged the right sysroot for the given settings.

This issues is relevant to the #5202 discussion

**Comment by Ignition** [comment]
The sysroot package is still a work in progress but ATM it sets some envs
```
def package_info(self):
    self.env_info.SYSROOT = self.package_folder
    self.env_info.CMAKE_SYSROOT = self.package_folder
```
This is working fine in most situations, but I've had autotools based projects where I've had to pass in sysroot explicitly.
```
env_build = AutoToolsBuildEnvironment(self)                         
sysroot = os.environ.get('SYSROOT')                                 
if sysroot:                                                         
    env_build.flags.append("--sysroot=%s" % sysroot)   
```

**Comment by Ignition** [comment]
The compiler poackage itself has `--with-sysroot` and `--with-build-sysroot` assigned. But still ironing out the kinks in trying to make the package work in all situations. 

E.g. package `gcc_installer` built on ubuntu, to be virtualenv installed onto another developers host machine which is rhel7 to produce target binaries for rhel6. This currently is not working but I sure this is down to some nuance in building gcc which is my fault and not a conan issue.

**Comment by Minimonium** [comment]
The distribution settings' moment is related to #3972

**Comment by memsharded** [comment]
BTW, the follow up of https://github.com/conan-io/conan/pull/5202 are:

- https://github.com/conan-io/conan/pull/5594
- https://github.com/conan-io/conan/pull/5592

The idea is that we will have a full "build" profile with all the necessary information for the build platform, that will fit more naturally than defining "xxx_build" settings as "os_build" and "arch_build".  But it is true that we are not considering to model the "target" context in the first iteration. 

cc /@jgsogo

**Comment by jgsogo** [comment]
Hi! As it is said before, we are working on a better approach for cross-compiling, but we are not entering (yet) the problem related to the `target` platform. We think we have enough for this first iteration with the `host` and `build` machines, although we write down every feedback related to it.

For the `target`, as you said, we will need a full profile to get/build libraries that should be provided by the `gcc` targetting the final distribution, but right now we are a little bit far from proposing a solution for it (there is quite interesting feedback in [these comments](https://github.com/conan-io/conan/pull/5202#issuecomment-507443426)). I know that @SSE4 has made some preliminary approaches to gcc-crosscompiler using the new feature, not even a POC, but feel free to ask him for advice on how to do it right now and how the future could look like.

**Comment by jgsogo** [comment]
https://github.com/conan-io/conan/pull/5594 is going to be merged in 1.24, it will allow defining two different profiles, one for the _host machine_ (binaries/libraries we are generating) and another one for the _build machine_ (tools we need to use).

In the use-case of a cross-compiler, that needs also the `target` information (which in this context is the _host profile_), the workaround after #5594 is to use an option in the _build profile_. That option will help to select the proper `gcc`, `binutils`, `sysroot` package (depending on the target RHEL6, RHEL7,...).

I would really like to know, @Ignition , if this approach alleviates your pain. We need to explore these scenarios to keep adding functionality that is needed for more complex graphs.

Thanks for your patience

**Comment by Ignition** [comment]
@jgsogo thanks, I'll see how the new approach alleviates certain work arounds I currently have in place and report what still the new solution looks like.
